src/macaron/repo_finder/repo_verifier.py

```diff
@@ -0,0 +1,300 @@
+        default_res = RepositoryVerificationResult(
+            status=RepositoryVerificationStatus.UNKNOWN, reason="unsupported_type", build_tool=self.build_tool
+        )
+
+        match self.build_tool:
+            case Maven():
+                git_ns_ver = self._verify_maven_git_ns()
+                if git_ns_ver.status == RepositoryVerificationStatus.PASSED:
+                    return git_ns_ver
+                return self._verify_maven()
+            case Gradle():
+                git_ns_ver = self._verify_maven_git_ns()
+                if git_ns_ver.status == RepositoryVerificationStatus.PASSED:
+                    return git_ns_ver
+                return self._verify_gradle()
+            # TODO: add verifiers for other types
+            case Poetry():
+                return default_res
+            case Pip():
+                return default_res
+            case Docker():
+                return default_res
+            case NPM():
+                return default_res
+            case Yarn():
+                return default_res
+            case Go():
+                return default_res
+            case _:
+                raise NotImplementedError(f"Unsupported build tool: {self.build_tool}")
+
+    def _same_group(self, g1: str, g2: str) -> bool:
+        if g1 == g2:
+            return True
+
+        g1_parts = g1.split(".")
+        g2_parts = g2.split(".")
+        if min(len(g1_parts), len(g2_parts)) < 2:
+            return False
+
+        if (g1_parts[0] in {"io", "com"} and g1_parts[1] in self._known_maven_namespaces) or (
+            g2_parts[0] in {"io", "com"} and g2_parts[1] in self._known_maven_namespaces
+        ):
+            if len(g1_parts) >= 3 and len(g2_parts) >= 3:
+                return all(g1_parts[i] == g2_parts[i] for i in range(3))
+            return False
+
+        for i in range(2):
+            if g1_parts[i] != g2_parts[i]:
+                return False
+
+        return True
+
+    @staticmethod
+    def _bfs_walk(root_dir: Path, filename: str) -> Path | None:
+        if not os.path.exists(root_dir) or not os.path.isdir(root_dir):
+            return None
+
+        queue: deque[Path] = deque()
+        queue.append(Path(root_dir))
+        while queue:
+            current_dir = queue.popleft()
+
+            # don't look through non-main directories
+            if any(
+                keyword in current_dir.name.lower()
+                for keyword in ["test", "example", "sample", "doc", "demo", "spec", "mock"]
+            ):
+                continue
+
+            if (current_dir / filename).exists():
+                return current_dir / filename
+
+            # ignore symlinks to prevent potential infinite loop
+            sub_dirs = [Path(it) for it in current_dir.iterdir() if it.is_dir() and not it.is_symlink()]
+            queue.extend(sub_dirs)
+
+        return None
+
+    def _verify_maven_git_ns(self) -> RepositoryVerificationResult:
```

Comment:
Wouldn't it make sense to add the Maven related functions to the [Maven build tool](https://github.com/oracle/macaron/blob/edfe06eb3f9337f86215b253f40f53ca1a0105c9/src/macaron/slsa_analyzer/build_tool/maven.py#L23) class and do the same later for the other build tools?

Reply:
Yeah actually I was thinking to have separate `RepoVerifier` classes for each build tool (probably all implementing a single abc or interface). But I thought that could be done when we actually have some implementation for other build tools. I can add that implementation now and see. Wdyt? 

However, I am not sure if it's a good idea to put repository verification logic directly inside a build tool class. I mean one would think it's too much of a (somewhat not related) responsibility for that class.

Reply:
Now each repo verifier has its own class. Let me know what you think